nano/secure/rep_weights.cpp

```diff
@@ -0,0 +1,139 @@
+#include <nano/secure/rep_weights.hpp>
+#include <nano/store/component.hpp>
+#include <nano/store/rep_weight.hpp>
+
+nano::rep_weights::rep_weights (nano::store::rep_weight & rep_weight_store_a) :
+	rep_weight_store{ rep_weight_store_a }
+{
+}
+
+void nano::rep_weights::representation_add (store::write_transaction const & txn_a, nano::account const & rep_a, nano::uint128_t const & amount_a)
+{
+	auto previous_weight{ rep_weight_store.get (txn_a, rep_a) };
+	auto new_weight = previous_weight + amount_a;
+	put_store (txn_a, rep_a, previous_weight, new_weight);
+	nano::lock_guard<nano::mutex> guard (mutex);
+	put_cache (rep_a, new_weight);
+}
+
+void nano::rep_weights::representation_add_dual (store::write_transaction const & txn_a, nano::account const & rep_1, nano::uint128_t const & amount_1, nano::account const & rep_2, nano::uint128_t const & amount_2)
+{
+	if (rep_1 != rep_2)
+	{
+		auto previous_weight_1{ rep_weight_store.get (txn_a, rep_1) };
+		auto previous_weight_2{ rep_weight_store.get (txn_a, rep_2) };
+		auto new_weight_1 = previous_weight_1 + amount_1;
+		auto new_weight_2 = previous_weight_2 + amount_2;
+		put_store (txn_a, rep_1, previous_weight_1, new_weight_1);
+		put_store (txn_a, rep_2, previous_weight_2, new_weight_2);
+		nano::lock_guard<nano::mutex> guard (mutex);
+		put_cache (rep_1, new_weight_1);
+		put_cache (rep_2, new_weight_2);
+	}
+	else
+	{
+		representation_add (txn_a, rep_1, amount_1 + amount_2);
+	}
+}
+
+void nano::rep_weights::representation_put (nano::account const & account_a, nano::uint128_t const & representation_a)
+{
+	nano::lock_guard<nano::mutex> guard (mutex);
+	put_cache (account_a, representation_a);
+}
+
+nano::uint128_t nano::rep_weights::representation_get (nano::account const & account_a) const
+{
+	nano::lock_guard<nano::mutex> lk (mutex);
+	return get (account_a);
+}
+
+/** Makes a copy */
+std::unordered_map<nano::account, nano::uint128_t> nano::rep_weights::get_rep_amounts () const
+{
+	nano::lock_guard<nano::mutex> guard (mutex);
+	return rep_amounts;
+}
+
+void nano::rep_weights::copy_from (nano::rep_weights & other_a)
+{
+	nano::lock_guard<nano::mutex> guard_this (mutex);
```

Comment:
Is this really needed? I see the existing implementation loads the weights in to a local variable and then copies them; I don't see why it can't load them in-place without the copy.